mypy/plugins/dataclasses.py

```diff
@@ -386,3 +395,100 @@ def _collect_field_args(expr: Expression) -> Tuple[bool, Dict[str, Expression]]:
             args[name] = arg
         return True, args
     return False, {}
+
+
+def asdict_callback(ctx: FunctionContext) -> Type:
+    positional_arg_types = ctx.arg_types[0]
+
+    if positional_arg_types:
+        dataclass_instance = positional_arg_types[0]
+        dataclass_instance = get_proper_type(dataclass_instance)
+        if isinstance(dataclass_instance, Instance):
+            info = dataclass_instance.type
+            if is_type_dataclass(info):
+                if len(ctx.arg_types) == 1:
+                    return _asdictify(ctx.api, ctx.context, dataclass_instance)
+                else:
+                    # We can't infer a more precise for calls where dict_factory is set.
+                    # At least for now, typeshed stubs for asdict don't allow you to pass in
+                    # `dict` as dict_factory, so we can't special-case that.
+                    return ctx.default_return_type
+
+    ctx.api.fail("'dataclasses.asdict' should be called on dataclass instances",
+                 ctx.context)
+    return ctx.default_return_type
+
+
+def _transform_type_args(*, typ: Instance, transform: Callable[[Instance], Type]) -> List[Type]:
```

Comment:
It seems to me all these dances with callbacks and two mutually recursive functions can be avoided (and the whole logic made more straightforward) with a small `TypeTransformVisitor`.

Reply:
I agree that it was too complex for what it did: I removed the function with the callback.
I think it's reasonable now, just one recursive function. I'm not sure how I would use `TypeTransformVisitor` here though. If you think it would help and could give some pointers, then I'd try and use it.

Reply:
Ah, I see it can be done with `TypeTranslator`. I've made the refactor. It does seem more straightforward once I understood how this TypeVisitor stuff worked.